python/tank/authentication/shotgun_shared/saml2_sso.py

```diff
@@ -0,0 +1,576 @@
+            return self._dialog.exec_()
+
+    def on_sso_login_cancel(self, event):
+        """
+        Called to cancel an ongoing login attempt.
+        """
+        log.debug("==- on_sso_login_cancel")
+
+        # We only need to cancel if there is login attempt currently being made.
+        if self.is_handling_event():
+            self.stop_session_renewal()
+            self.resolve_event(end_session=True)
+        self._dialog.accept()
+
+    def on_dialog_closed(self, result):
+        """
+        Called whenever the dialog is dismissed.
+
+        This can be the result of a callback, a timeout or user interaction.
+        """
+        log.debug("==- on_dialog_closed")
+
+        if self.is_handling_event():
+            if result == QtGui.QDialog.Rejected and self._session.cookies != "":
+                # We got here because of a timeout attempting a GUI-less login.
+                # Let's clear the cookies, and force the use of the GUI.
+                self._session.cookies = ""
+                # Let's have another go, without any cookies this time !
+                self.on_sso_login_attempt()
+            else:
+                # end_session = result == QtGui.QDialog.Rejected
+                # self.resolve_event(end_session=end_session)
+                self.resolve_event()
+        else:
+            # Should we get a rejected dialog, then we have had a timeout.
+            if result == QtGui.QDialog.Rejected:
+                # @FIXME: Figure out exactly what to do when we have a timeout.
+                log.warn("Our QDialog got canceled outside of an event handling...")
+
+        # Clear the web page
+        self._view.page().mainFrame().load("about:blank")
+
+    def on_sso_enable_renewal(self, event):
+        """
+        Called when enabling automatic SSO session renewal.
+
+        A new session will be created if there is not already a current one.
+        This will be in the case of the automatic (and successful)
+        authentication at the startup of the application.
+
+        """
+        log.debug("==- on_sso_enable_renewal")
+
+        contents = json.loads(event.contents())
+
+        if self._session is None:
+            self.start_new_session({
+                "host": contents["params"]["site_url"],
+                "cookies": contents["params"]["cookies"]
+            })
+        self.start_sso_renewal()
+
+    def on_sso_disable_renewal(self, event):
+        """
+        Called to disable automatic session renewal.
+
+        This will be required when switching to a new connection (where the new
+        site may not using SSO) or at the close of the application.
+        """
+        log.debug("==- on_sso_disable_renewal")
+
+        self.stop_session_renewal()
+
+################################################################################
+#
+# functions
+#
+################################################################################
+
+
+def _decode_cookies(rv_cookies):
```

Comment:
This method is actually used elsewhere. The variable name is simply unfortunate.